File: wrappers/nodejs/types/utils/vault.ts
import { Credential, CredentialEntry, CredentialMetadata, Vault } from "../..";

class WrappedVault implements Vault {
  constructor(private readonly vault: Vault) {
    this.storeCredential = this.storeCredential.bind(this);
    this.deleteCredential = this.deleteCredential.bind(this);
    this.findCredentials = this.findCredentials.bind(this);
    this.getCredential = this.getCredential.bind(this);
    this.getCredentials = this.getCredentials.bind(this);
    this.findCredentials = this.findCredentials.bind(this);
  }

  async storeCredential(credential: Credential, metadata: CredentialMetadata): Promise<string> {
    return await this.vault.storeCredential(credential, metadata);
  }

  async deleteCredential(id: string): Promise<void> {
    return await this.vault.deleteCredential(id);
  }

  async findCredentials(criteria: Array<string>): Promise<Array<CredentialEntry>> {
    return await this.vault.findCredentials(criteria);
  }

  async getCredential(id: string): Promise<CredentialEntry | null> {
    return await this.vault.getCredential(id);
  }

  async getCredentials(): Promise<Array<CredentialEntry>> {
    return await this.vault.getCredentials();
  }
}

export function contextEnsuredVault(vault: Vault): Vault {
  return new WrappedVault(vault);
}

File: wrappers/uniffi/src/inmem/vault.rs
use async_trait::async_trait;
use equs_sdk::vault::Vault as EqusSdkVault;
use equs_sdk::vc::{Credential, CredentialMetadata};

use crate::common::{Error, Result};
use crate::vault::{CredentialEntry, Vault, VaultFetchOptions};

#[derive(uniffi::Object, Debug)]
pub struct InMemVault(equs_sdk::inmem::vault::InMemVault);

#[uniffi::export()]
#[async_trait]
impl Vault for InMemVault {
    async fn store_credential(
        &self,
        credential: Credential,
        metadata: CredentialMetadata,
    ) -> Result<String> {
        self.0
            .store_credential(credential, &metadata)
            .await
            .map_err(|err| Error::Vault(format!("{:?}", err)))
    }

    async fn delete_credential(&self, id: String) -> Result<()> {
        self.0
            .delete_credential(&id)
            .await
            .map_err(|err| Error::Vault(format!("{:?}", err)))
    }

    async fn get_credential(&self, id: String) -> Result<Option<CredentialEntry>> {
        self.0
            .get_credential(&id)
            .await
            .map_err(|err| Error::Vault(format!("{:?}", err)))
    }

    async fn get_credentials(
        &self,
        options: Option<VaultFetchOptions>,
    ) -> Result<Vec<CredentialEntry>> {
        let options = options.map(Into::into);

        self.0
            .get_credentials(options)
            .await
            .map_err(|err| Error::Vault(format!("{:?}", err)))
    }

    async fn find_credentials(
        &self,
        fields: Vec<String>,
        options: Option<VaultFetchOptions>,
    ) -> Result<Vec<CredentialEntry>> {
        let options = options.map(Into::into);

        self.0
            .find_credentials(fields, options)
            .await
            .map_err(|err| Error::Vault(format!("{:?}", err)))
    }
}

#[uniffi::export()]
impl InMemVault {
    #[uniffi::constructor]
    fn new() -> Self {
        InMemVault(equs_sdk::inmem::vault::InMemVault::new())
    }
}

impl InMemVault {
    pub fn inner(&self) -> equs_sdk::inmem::vault::InMemVault {
        self.0.clone()
    }
}
